cpp/src/strings/like.cu

```diff
@@ -0,0 +1,152 @@
+  string_view const d_pattern;
+  string_view const d_escape;
+
+  __device__ bool operator()(size_type const idx)
+  {
+    if (d_strings.is_null(idx)) return false;
+    auto const d_str = d_strings.element<string_view>(idx);
+
+    // using only iterators to better handle UTF-8 characters
+    auto target_itr  = d_str.begin();
+    auto pattern_itr = d_pattern.begin();
+
+    auto const target_end  = d_str.end();
+    auto const pattern_end = d_pattern.end();
+    auto const esc_char    = d_escape.empty() ? 0 : d_escape[0];
+
+    auto last_target_itr  = target_end;
+    auto last_pattern_itr = pattern_end;
+
+    bool result = true;
+    while (true) {
+      // walk through the pattern and check against the current character
+      while (pattern_itr < pattern_end) {
+        auto const escaped = *pattern_itr == esc_char;
+        auto const ptn_ch =
+          escaped && (pattern_itr + 1 < pattern_end) ? *(++pattern_itr) : *pattern_itr;
+
+        if (escaped || (ptn_ch != multi_wildcard)) {
+          // check match with the current character
+          result = ((target_itr != target_end) &&
+                    ((!escaped && ptn_ch == single_wildcard) || (ptn_ch == *target_itr)));
+          if (!result) { break; }
+          ++target_itr;
+          ++pattern_itr;
+        } else {
+          // process wildcard '%'
+          result = true;
+          ++pattern_itr;
+          if (pattern_itr == pattern_end) {  // pattern ends with '%' so we are done
+            target_itr = target_end;
+            break;
+          }
+          // save positions
+          last_pattern_itr = pattern_itr;
+          last_target_itr  = target_itr;
+        }
+      }  // next pattern character
+
+      if (result && (target_itr == target_end)) { break; }  // success
+
+      result = false;
+      // check if exhausted either the pattern or the target string
+      if (last_pattern_itr == pattern_end || last_target_itr == target_end) { break; }
+
+      // restore saved positions
+      pattern_itr = last_pattern_itr;
+      target_itr  = ++last_target_itr;
+    }
+    return result;
+  }
+};
+
+}  // namespace
+
+std::unique_ptr<column> like(
+  strings_column_view const& input,
+  string_scalar const& pattern,
+  string_scalar const& escape_character,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+{
+  auto results = make_numeric_column(data_type{type_id::BOOL8},
+                                     input.size(),
+                                     cudf::detail::copy_bitmask(input.parent(), stream, mr),
+                                     input.null_count(),
+                                     stream,
+                                     mr);
+  if (input.is_empty()) { return results; }
+
+  CUDF_EXPECTS(pattern.is_valid(stream), "Parameter pattern must be valid");
+  CUDF_EXPECTS(escape_character.is_valid(stream), "Parameter escape_character must be valid");
```

Comment:
I would rather not since the string is in device memory and there is no easy way to check for a single character (which can be multiple bytes) without launching a kernel.

Reply:
Can it be documented that the string must be a single character, and validate the length in the Python API?

Reply:
Reference issue #5505